asynction/security.py

```diff
@@ -0,0 +1,367 @@
+        raise SecurityException("Missing Bearer info function")
+    try:
+        return load_handler(scheme.x_bearer_info_func)
+    except ValueError as err:
+        raise SecurityException from err
+
+
+def validate_token_info(
+    token_info: InternalSecurityCheckResponse,
+    scope_validate_func: ScopeValidateFunc,
+    required_scopes: Sequence[str],
+) -> InternalSecurityCheckResponse:
+    scopes: Optional[Sequence[str]] = None
+    if not token_info:
+        return None
+
+    if "scopes" in token_info:
+        scopes = token_info.get("scopes")
+    elif "scope" in token_info:
+        scope = token_info.get("scope")
+        if isinstance(scope, str):
+            scopes = scope.split()
+        else:
+            raise ValueError("'scope' should be a string")
+
+    if not scopes:
+        raise ValueError("missing scopes in token info")
+
+    if not scope_validate_func(required_scopes, scopes):
+        raise SecurityException(
+            f"Invalid scopes: required: {required_scopes}, provided: {scopes}"
+        )
+
+    return token_info
+
+
+def build_http_security_check(
+    requirement: InternalSecurityRequirement, scheme: SecurityScheme
+) -> Optional[InternalSecurityCheck]:
+    _, required_scopes = requirement
+
+    if scheme.scheme == HTTPAuthenticationScheme.BASIC:
+        basic_info_func = load_basic_info_func(scheme)
+
+        def http_security_check(request: Request) -> InternalSecurityCheckResponse:
+            return validate_basic(request, basic_info_func, required_scopes)
+
+        return http_security_check
+    elif scheme.scheme == HTTPAuthenticationScheme.BEARER:
+        bearer_info_func = load_bearer_info_func(scheme)
+        bearer_format = scheme.bearer_format
+
+        def http_bearer_security_check(
+            request: Request,
+        ) -> InternalSecurityCheckResponse:
+            return validate_bearer(
+                request, bearer_info_func, required_scopes, bearer_format
+            )
+
+        return http_bearer_security_check
+    else:
+        return None
+
+
+def build_http_api_key_security_check(
+    requirement: InternalSecurityRequirement, scheme: SecurityScheme
+) -> Optional[InternalSecurityCheck]:
+    api_key_info_func = load_api_key_info_func(scheme)
+    _, required_scopes = requirement
+
+    def http_api_key_security_check(request: Request) -> InternalSecurityCheckResponse:
+        api_key = (
+            (
+                {
+                    ApiKeyLocation.QUERY: request.args,
+                    ApiKeyLocation.HEADER: request.headers,
+                    ApiKeyLocation.COOKIE: request.cookies,
+                }
+                .get(scheme.in_, {})
+                .get(scheme.name)
+            )
+            # NOTE: this is technically not needed, but mypy insists it is checked
+            # because if is unaware that we have validated it previously
+            if scheme.in_ and scheme.name
+            else None
+        )
```

Comment:
```suggestion
        assert scheme.in_ and scheme.name  # to make mypy happy
        request_dicts = {
            ApiKeyLocation.QUERY: request.args,
            ApiKeyLocation.HEADER: request.headers,
            ApiKeyLocation.COOKIE: request.cookies,
        }
        try:
            return requests_dicts[scheme.in_][scheme.name]
        except KeyError:
            return None
```
Just making it a bit more readable. Not a strong opinion on this one.